src/script/interpreter.cpp

```diff
@@ -1433,6 +1513,9 @@ void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut>&& spent
     // Determine which precomputation-impacting features this transaction uses.
     bool uses_bip143_segwit = false;
     bool uses_bip341_taproot = false;
+    // TODO: Improve this heuristic
+    bool uses_bip119_ctv = true;
```

Comment:
As the TODO says, should be improved

Reply:
yes; this became a TODO post taproot because we started doing much more aggressive/expensive heuristics, scanning to see if it's a taproot spend. We used to just IIRC always compute all caches.


It's not clear to me the performance benefit of scanning v.s. always computing the hashes @sipa. 50% of transactions are SegWit these days (https://charts.woobull.com/bitcoin-segwit-adoption/), and CTV only does extra non O(1) hashing in the event that it's a transaction that uses non segwit inputs.

It'd be also possible to just *never* cache CTV, if the extra work seems not worth it. Or to only cache for single input txns, only cache for txns with a bare script standard CTV, only cache for segwit or has bare CTV, etc. It's kinda hard to say what the optimal caching policy is and it's bikesheddable + non consensus, so I chose to just leave it for now as the extra work is not terrible, and more advanced policies open up the potential for quadratic hashing bugs (e.g., txn has a scriptSig on every input except the CTV one and the CTV one is <H> CTV CTV... and we don't cache).

Reply:
One last thing on caching: CTV was designed so that if you're expecting to operate the same CTV at diff input indexes, you can cache the sha256 midstate because index is the final field. However, I think that it's relatively rare you'd want CTV scripts at indexes other than 0.

Reply:
the main reason to cache the hashing it to prevent scripts like `<H> CTV CTV CTV.... CTV` from being able to cause N^2 hashing to occur. Fixed hashing (like if the input is at input one, not the 'usual' 0), of e.g. 80 bytes (idk offhand how long ctv digest is) is not an issue as it's like script: `<v> HASH256 HASH256 HASH256 HASH256...`, but it is a problem if you have to hash an entire (big) tx.

hence the mention of quadratic hashing above. Caching is required for "safety", not performance.

Reply:
> the main reason to cache the hashing it to prevent scripts like <H> CTV CTV CTV.... CTV from being able to cause N^2 hashing to occur.

@JeremyRubin what's the reason that we can't populate a cache "inline" while we're doing script execution? That way we could both avoid having to precompute for transactions that don't use CTV and avoid N^2 hashing. I assume there's some limitation that prevented you from doing this in the first place.

Reply:
This is the approach now taken in this PR!